### ISSUE TITLE
Add icdf functions for distributions

### ISSUE THREAD

**Issue by michaelraczycki** [pymc-devs/pymc#6612]
### Description

We are looking for help to implement inverse cumulative distribution (ICDF) functions for our distributions!

### How to help?

This PR should give a template on how to implement and test new icdf functions for distributions: https://github.com/pymc-devs/pymc/pull/6528

ICDF functions allow users to get the value associated with a specific cumulative probability. 

So far we've added 2 examples for continuous distribution 
* Uniform:
https://github.com/pymc-devs/pymc/blob/2fcce433548a411f695d684ffb2d001a82a35f20/pymc/distributions/continuous.py#L348-L351
* Normal: 
https://github.com/pymc-devs/pymc/blob/2fcce433548a411f695d684ffb2d001a82a35f20/pymc/distributions/continuous.py#L541-L548

And an example for a discrete distribution:
* Geometric:
https://github.com/pymc-devs/pymc/blob/2fcce433548a411f695d684ffb2d001a82a35f20/pymc/distributions/discrete.py#L824-L832

Multiple sources describing the icdf function for any specific distribution can be found, you're free to choose which one is working for you. To start with I recommend checking:

* https://help.imsl.com/c/2016/html/cnlstat/index.html#page/CNL%2520Stat%2Fcsch11.14.01.html%23
* Wikipedia,
E.g.: https://en.wikipedia.org/wiki/Normal_distribution
 ![Zrzut ekranu 2023-03-18 o 15 49 21](https://user-images.githubusercontent.com/119355382/226107058-645d110a-07ad-445d-b540-b8b73b6346ef.png)
 It should be called "Quantile" as on the screenshot above.

New tests have to be added in test_continuous.py for continuous distributions, and test_discrete.py for discrete ones. You can use existing tests as a template:

https://github.com/pymc-devs/pymc/blob/2fcce433548a411f695d684ffb2d001a82a35f20/tests/distributions/test_continuous.py#L282-L286

Don't hesitate to ask any questions. You can grab as many distributions to implement moments as you want. Just make sure to write in this issue so that we can keep track of it.

Profit with your new open source KARMA!

The following distributions don't have an icdf method implemented:

- [x]  Beta #6845
- [x]  Kumaraswamy  #6642 
- [x]  Exponential #6641
- [x]  Laplace #6707 
- [x]  StudentT #6845 
- [x]  Cauchy  #6747
- [x]  HalfCauchy
- [x]  Gamma #6845 
- [x]  HalfNormal
- [x]  Weibull  #6802
- [x]  LogNormal  #6766
- [x]  HalfStudentT
- [ ]  Wald
- [x]  Pareto #6707 
- [x]  InverseGamma
- [ ]  ExGaussian
- [ ]  Binomial  #7362 
- [ ]  BetaBinomial
- [ ]  Poisson
- [ ]  NegativeBinomial
- [ ]  DiracDelta
- [x]  DiscreteUniform #6617 
- [ ]  HyperGeometric
- [ ]  Categorical
- [ ]  CustomDist (allow user to pass one, or try to infer like we do for logp/logcdf already)
- [ ]  AsymmetricLaplace
- [ ]  SkewNormal
- [x]  Triangular  #6802
- [ ]  DiscreteWeibull
- [x]  Gumbel  #6802
- [x]  Logistic  #6747
- [x]  LogitNormal
- [ ]  Interpolated
- [ ]  Rice
- [x]  Moyal  #6802 
- [ ]  PolyaGamma
- [ ] Mixture (requires an interative algorithm based on the logcdf, see [here](https://stats.stackexchange.com/questions/14481/quantiles-from-the-combination-of-normal-distributions))

Note that not all of the icdf equations will have closed solution, so it's recommended to first start with the ones that can be found in closed form, as they will be easier to implement and will contribute to the task further with providing other contributors with templates to understand the topic better. The list above is not final, and I'll try to update it to contain all distributions available for taking.

**Comment by gokuld** [comment]
Hi @michaelraczycki, or whoever is concerned, this is my first attempt to contribute to PyMC, so please let me know if this PR makes sense.
If this turns out to be useful I am happy to add ICDF functions for more distributions.

**Comment by michaelraczycki** [comment]
Hey @gokuld! Thank you for your contribution, it look promising.  For the future reference please add a comment under the issue, letting others  know that you're starting to work on specific issue/ parts of the issue. This assure that you're not working in parallel with someone else on the same part of the development.

**Comment by gokuld** [comment]
> Hey @gokuld! Thank you for your contribution, it look promising. 

Thanks @michaelraczycki .

> For the future reference please add a comment under the issue, letting others know that you're starting to work on specific issue/ parts of the issue. This assure that you're not working in parallel with someone else on the same part of the development.

Sure, I will post a comment when I start work to avoid parallel duplicate work from next time!

**Comment by gokuld** [comment]
Hey all, I am starting work on the ICDF for the continuous beta distribution.  Let me know if anyone else is working on this already.
(@michaelraczycki)

**Comment by michaelraczycki** [comment]
@gokuld if there's no comment under this issue saying that someone reserves it you don't need to ask. Just call it and it's yours :)
Also in case so for any reason you can't / don't want to work on the issue anymore please also let us know here. 
Good luck!

**Comment by ricardoV94** [comment]
@gokuld AFAICT the beta distribution doesn't have a closed form solution, so you would need an iterative algorithm which may not be trivial to write if you are not familiar with PyTensor. Ignore if you were aware of the fact ;)

**Comment by james-2001** [comment]
I'll be picking up
- Laplace 
- Pareto

Cheers :smile:

**Comment by gokuld** [comment]
I have no immediate plans of finishing the work on the beta distribution, anyone else interested may pick it up.

**Comment by michaelraczycki** [comment]
Good luck @james-2001 , and than you for your contribution @gokuld !

**Comment by amyoshino** [comment]
I will work on the LogNormal :)

**Comment by amyoshino** [comment]
I'm tackling now:
- Cauchy distribution
- Logistic distribution 

😄

**Comment by ricardoV94** [comment]
This should suffice for all the "Half"Distributions:

```
def icdf(value, *args)
  return icdf(Full.dist(*args), value)
```
Where `Full` is the equivalent non-half version of the distribution (Normal for HalfNormal, Student for HalfStudent and so on).

**Comment by amyoshino** [comment]
Thanks for the tips, I am going to work on and add to the next PR the `Half Normal` implementation since I am working on the lognormal already. Once I get used to this new approach I can try to adapt it to the other "Half".

**Comment by ricardoV94** [comment]
Thanks for checking out the half idea @amyoshino, I realized these don't work as other transforms and the automatic icdf should raise. I opened a PR for that effect: https://github.com/pymc-devs/pymc/pull/6793

**Comment by amyoshino** [comment]
> Thanks for checking out the half-dist idea @amyoshino, you made me realize these don't work as other transforms and the automatic icdf should raise. I opened a PR for that effect: #6793

I'm glad I was able to help! 😄

**Comment by amyoshino** [comment]
I will now work on the Triangular distribution :)

**Comment by amyoshino** [comment]
It looks like the remaining ones have no closed form (not that I have found so far). I will give it a try on developing the icdf functions for the remaining ones. It might take a while but I will do my best to get used to all we need as fast as possible. 
So, just to get some focus in here, I will start with some that require the Inverse Regularized Gamma function and Inverse Regularized Beta function to get used to its implementation: 
- Gamma Distribution
- ChiSquared Distribution
- Beta Distribution
- StudentT Distribution

**Comment by niknow** [comment]
I will work on Binomial now.

**Comment by fireddd** [comment]
Hi, are there any more functions that need to be added here?

**Comment by ricardoV94** [comment]
> Hi, are there any more functions that need to be added here?

I'm not 100% if the list is up to date, but it suggests missing distributions

**Comment by berkalpay** [comment]
I'll work on the InverseGamma

**Comment by iTzZe4lous** [comment]
I can work on Categorical if its yet to be done

**Comment by ricardoV94** [comment]
> I can work on Categorical if its yet to be done

If it's not yet, just go ahead. The way to claim work is to open a PR :)

**Comment by dalai444** [comment]
I'll work on the HalfStudentT

**Comment by darmendariz** [comment]
Hello all. I'm taking a stab at the icdf for AsymmetricLaplace. I see a [closed PR](https://github.com/pymc-devs/pymc/pull/7141) from almost 2 years ago with unmerged commits, so I'm assuming it's free to take on.

**Comment by Youssef-Naggar** [comment]
Hello world. 
I would like to tackle the ICDF implementation for the DiscreteWeibull and DiracDelta distributions.
I have the code ready and am currently finalising the tests. I will open a PR shortly!

**Comment by skysabarre** [comment]
I'd like to implement the icdf for HalfNormal. Starting on it now.

**Comment by skysabarre** [comment]
Update: I ended up contributing to issue #6408 instead with PR #8224. Still interested in contributing further to #6612 if any distributions remain unclaimed.

**Comment by jacobpstein** [comment]
Hi! I'd like to pick up some of the missing `icdf` methods. This has come up in recent work of mine (hierarchical Bernoulli models where I needed quantiles for calibration and inverse-transform sampling), so I'm motivated to help close the gap.

Starting with the closed-form cases that are still missing, which look like clean first additions:

- **`Bernoulli`** — `icdf(q) = 0 if q ≤ 1 − p else 1`
- **`AsymmetricLaplace`** — piecewise closed-form quantile

I'll follow the existing convention (`check_icdf_value` / `check_icdf_parameters`) and add tests against the SciPy `ppf` using the `check_icdf` helper.

One thing worth flagging: the count distributions (`Poisson`, `Binomial`, `NegativeBinomial`) don't have closed-form quantile functions, and I don't see a numerical-inverse fallback in the codebase — `gammaincinv`/`betaincinv` invert the continuous CDFs rather than the integer quantiles. Supporting those would need either a bisection/search approach or a SciPy `Op`. Happy to take that on separately if there's interest — wanted to check the preferred direction before going down that path.

Will open a draft PR for `Bernoulli` + `AsymmetricLaplace` shortly.

**Comment by jacobpstein** [comment]
Quick correction after checking existing work: **AsymmetricLaplace** is already covered by #7976 (and claimed by @darmendariz above), so I'll drop it from my scope — apologies for the overlap.

I'll instead take on:

- **Bernoulli** — closed form, still missing (`icdf(q) = 0 if q ≤ 1 − p else 1`). It's not on the checklist above, but `pm.icdf` currently raises `NotImplementedError` for it.
- The **count distributions** (`Poisson`, `NegativeBinomial`, `Binomial`) — these have no closed-form quantile, and the special-function inverses (`gammaincinv`/`betaincinv`) don't apply, since they invert the CDF in its continuous probability argument rather than over the integer support. So I'll implement an iterative `logcdf`-based search (the same approach already noted for `Mixture`) via a shared helper.

I'll open a draft PR so the approach can be reviewed concretely — happy to adjust direction, e.g. how to bracket the search for the unbounded-support cases. (For reference, the earlier Binomial attempt in #7362 was closed unmerged.)

**Comment by prisharai** [comment]
TruncatedNormal isn't in the original list but currently has logcdf without icdf; pm.icdf raises NotImplementedError for it. It has a closed-form inverse, so I'd like to contribute it.

**Comment by TalenMud** [comment]
Hi, I'd like to implement the icdf for the Wald distribution. I'll follow the existing pattern in `continuous.py` and open a PR shortly.

**Comment by Sharman1661** [comment]
Hello, is this list up to date?
If so, I'll begin work on the Asymmetric Laplace and the Dirac-Delta distributions.
